### ISSUE TITLE
Roll out KG2.8.0.1c

### ISSUE THREAD

**Issue by amykglen** [RTXteam/RTX#2031]
_NOTE: To create a new issue based on this template, simply go to: https://github.com/RTXteam/RTX/issues/new?template=kg2rollout.md_

**THE BRANCH FOR THIS ROLLOUT IS: `kg2.8.0.1c`**
**THE ARAX-DATABASES.RTX.AI DIRECTORY FOR THIS ROLLOUT IS: `/home/rtxconfig/KG2.8.0.1`**

#### 1. Build and load KG2c:

- [x] merge `master` into the branch being used for this KG2 version
- [x] update the four hardcoded biolink version numbers in the branch (as needed):
  - [x] in `code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml` ([github](https://github.com/RTXteam/RTX/tree/master/code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml#L18); [local](../code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml))
  - [x] in `code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml` ([github](https://github.com/RTXteam/RTX/tree/master/code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml#L18); [local](../code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml))
  - [x] in `code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml` ([github](https://github.com/RTXteam/RTX/tree/master/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml#L17); [local](../code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml))
  - [x] in `code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml` ([github](https://github.com/RTXteam/RTX/tree/master/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml#L17); [local](../code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml))
- [x] build a new KG2c on `buildkg2c.rtx.ai` from the branch (how-to is [here](https://github.com/RTXteam/RTX/tree/master/code/kg2c#build-kg2canonicalized))
  - [x] before starting the build:
    - [x] make sure there is enough disk space available on `arax-databases.rtx.ai` (need at least 100G, ideally >120G). delete old KG2 database directories as needed (warn the team on Slack in advance).
    - [x] make sure to choose to build a new synonymizer in `kg2c_config.json`, as described in the how-to
  - [x] after the build is done, verify it looks ok:
    - [x] `node_synonymizer.sqlite` should be around 8-15 GB
    - [x] make sure `node_synonymizer.sqlite`'s last modified date is today (or whatever day the build was run)
    - [x] make sure `kg2c_lite.json.gz`'s last modified date is today (or whatever day the build was run)
    - [x] the entire build runtime (synonymizer + KG2c) shouldn't have been more than 24 hours
    - [x] the synonymizer and KG2c artifacts should have been auto-uploaded into the proper directory on `arax-databases.rtx.ai`
- [x] load the new KG2c into neo4j at http://kg2-X-Yc.rtx.ai:7474/browser/ (how to is [here](https://github.com/RTXteam/RTX/tree/master/code/kg2c#host-kg2canonicalized-in-neo4j))
  - [x] verify the correct KG2 version was uploaded by running this query: `match (n {id:"RTX:KG2c"}) return n`
- [x] update `RTX/code/config_dbs.json` in the branch:
  - [x] update the synonymizer version number/path
  - [x] update the fda_approved_drugs version number/path
  - [x] update the autocomplete version number/path
  - [x] update the meta_kg version number/path
  - [x] update the kg2c sqlite version number/path
  - [x] update the KG2pre and KG2c Neo4j endpoints
- [x] upload the new `kg2c_lite_2.X.Y.json.gz` file to the [translator-lfs-artifacts](https://github.com/ncats/translator-lfs-artifacts/tree/main/files) repo
- [x] upload the new `kg2_nodes_not_in_sri_nn.tsv` file to the [translator-lfs-artifacts](https://github.com/ncats/translator-lfs-artifacts/tree/main/files) repo
- [x] load the new KG2c into Plover (available at http://kg2cplover.rtx.ai:9990)
  - [x] in the Plover repo, update `kg_config.json` to point to the new version of the kg2c lite file (in the git lfs repo); push this change to a branch
  - [x] make any other changes to Plover code that this KG2 version necessitates (in that same branch)
  - [x] build a Plover from the branch
  - [x] update `config_dbs.json` in the RTX repo to point to this new Plover **for the 'dev' maturity level**

#### 2. Rebuild downstream databases:

The following databases should be rebuilt and copies of them should be put in `/home/rtxconfig/KG2.X.Y` on `arax-databases.rtx.ai`. Please use this kind of naming format: `mydatabase_v1.0_KG2.X.Y.sqlite`.

- [x] NGD database
- [x] refreshed DTD @chunyuma
- [x] DTD model @chunyuma _(may be skipped - depends on the changes in this KG2 version)
- [x] DTD database @chunyuma _(may be skipped - depends on the changes in this KG2 version)
- [ ] XDTD database @chunyuma

**NOTE**: As databases are rebuilt, `RTX/code/config_dbs.json` will need to be updated to point to their new paths! Push these changes to the branch for this KG2 version, unless the rollout of this KG2 version has already occurred, in which case you should push to `master` (but first follow the steps described [here](https://github.com/RTXteam/RTX/wiki/Config,-databases,-and-SFTP#config_dbsjson)). 

#### 3. Update the ARAX codebase:

All code changes should **go in the branch for this KG2 version**!

- [x] regenerate the KG2c test triples file in the branch for this KG2 version
  - [x] ensure the new KG2c Neo4j is currently running
  - [x] check out the branch and pull to get the latest changes (this is important for ensuring the correct KG2c Neo4j is used)
  - [x] run [create_json_of_kp_predicate_triples.py](https://github.com/RTXteam/RTX/blob/kg2.8.0.1c/code/ARAX/KnowledgeSources/create_json_of_kp_predicate_triples.py)
  - [x] push the regenerated file to `RTX/code/ARAX/KnowledgeSources/RTX_KG2c_test_triples.json`
- [x] update Expand code as needed
- [x] update any other modules as needed
- [x] test everything together:
  - [x] check out the branch and pull to get the latest changes
  - [x] locally set `force_local = True` in `ARAX_expander.py` (to avoid using the old KG2 API)
  - [x] then run the entire ARAX pytest suite
  - [x] address any failing tests
- [x] update the KG2 and ARAX version numbers in the appropriate places (`openapi.yaml`, etc.):
  - [x] Bump version for `RTX/code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml` ([github](https://github.com/RTXteam/RTX/blob/master/code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml#L12); [local](../code/UI/OpenAPI/python-flask-server/openapi_server/openapi/openapi.yaml)) in line 12 (`version:`); the major and minor release numbers are kept synchronous with the TRAPI version; just bump the patch release version (least significant digit)
  - [x] Bump version for `RTX/code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml` ([github](https://github.com/RTXteam/RTX/blob/master/code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml#L12); [local](../code/UI/OpenAPI/python-flask-server/KG2/openapi_server/openapi/openapi.yaml)) in line 12 (`version:`); the first three digits are kept synchronous with the KG2 release version
  - [x] Bump version number in `RTX/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml` ([github](https://github.com/RTXteam/RTX/blob/master/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml#L4); [local](../code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_ARAX.yaml)) on line 4 (`version:`); same as for the ARAX `openapi.yaml` file
  - [x] Bump version number in `RTX/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml` ([github](https://github.com/RTXteam/RTX/blob/master/code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml#L4); [local](../code/UI/OpenAPI/python-flask-server/RTX_OA3_TRAPI1.4_KG2.yaml)) on line 4 (`version:`); same as for the KG2 `openapi.yaml` file
  

#### 4. Pre-upload databases:

Before rolling out, we need to pre-upload the new databases (referenced in `config_dbs.json`) to `arax.ncats.io` and the ITRB SFTP server. These steps can be done well in advance of the rollout; it doesn't hurt anything to do them early.

- [x] make sure `arax.ncats.io` has at least 100G of disk space free; delete old KG2 databases to free up space as needed (warn the team on slack first)
- [x] `scp` the new databases from `arax-databases.rtx.ai` to `arax.ncats.io:/data/orangeboard/databases/KG2.X.Y`
- [x] upload the new databases and their md5 checksums to ITRB's SFTP server using the steps detailed [here](https://github.com/RTXteam/RTX/wiki/Config,-databases,-and-SFTP#steps-for-all-databases-at-once)

#### 5. Do the rollout:

- [x] merge `master` into the branch for this KG2 version
- [x] merge the branch into `master`
- [x] roll `master` out to the various `arax.ncats.io` endpoints
- [x] run the database manager
- [x] run the pytest suite on the various endpoints
- [x] verify the new KG2 version is actually being used
  - [x] run this JSON query in the ARAX/KG2 UIs: `{"nodes": {"n00": {"ids": ["RTX:KG2c"]}}, "edges": {}}`
  - [x] look up `RTX:KG2` in the Synonyms tab in the UI
- [x] update our CI/CD testing instance with the new databases:
  - [x] `ssh ubuntu@cicd.rtx.ai`
  - [x] `cd RTX`
  - [x] `git pull origin master`
  - [x] `python3 code/ARAX/ARAXQuery/ARAX_database_manager.py --mnt --skip-if-exists --remove_unused`

#### 6. Final items/clean up:

- [x] turn off the old KG2c version's neo4j instance
- [x] turn off the old KG2c version's plover instance
- [x] turn off the new KG2pre version's neo4j instance
- [ ] upgrade the ITRB Plover endpoint (https://kg2cploverdb.ci.transltr.io) to this KG2 version and make the KG2 API start using it (instead of our self-hosted endpoint): 
    - [x] update `kg_config.json` in the `main` branch of the Plover repo to point to the new `kg2c_lite_2.X.Y.json.gz` file (push this change)
    - [x] wait about 45 minutes for the endpoint to rebuild and then run Plover tests to verify it's working
    - [x] run the ARAX pytest suite with the NCATS endpoint plugged in (locally change the URL in `config_dbs.json` and set `force_local = True` in Expand)
    - [x] if all tests pass, update `config_dbs.json` in `master` to point to the ITRB Plover endpoints (all maturity levels)
    - [x] roll `master` out to the various endpoints on arax.ncats.io
    - [ ] turn off our plover endpoint and verify once more that ARAX is still working ok

**Comment by amykglen** [comment]
KG2.8.0.1c build is underway. I'll let Chunyu know when it's finished and ready for him to start rebuilding downstream databases (should be by tomorrow morning)

**Comment by amykglen** [comment]
Hey @chunyuma - so KG2.8.0.1c is ready for you to start building databases. It's loaded into neo4j at http://kg2-8-0-1c.rtx.ai:7474/browser/ and related files are in the `/KG2.8.0.1` directory on arax-databases.rtx.ai.

Please note a couple significant changes with this build: 
1. The `knowledge_source` property on edges has been renamed `primary_knowledge_source` and is now a string instead of a list.
2. Edges that have the same subject, predicate, and object are **no longer merged** _if they have different primary_knowledge_sources_. Meaning, there can be multiple edges in KG2c that have identical subject/predicate/object, but different primary knowledge sources. As part of this, the `id`s of KG2c edges now includes the `primary_knowledge_source`.

Please put any code changes into the `kg2.8.0.1c` branch. Let me know if you have any questions!

**Comment by chunyuma** [comment]
Hi @amykglen, I would like to give you some updates. I've refreshed the DTD databases and have uploaded them to the `arax-databases` server. I'm now re-building the xDTD model and databases for kg2.8.0.1c, which might need at least two weeks.  The model re-training must be completed before the relay, but pre-computed results may take an additional week to finish. Therefore, it is uncertain if all of them can be completed before the relay.

However, I'm modifying the code to make xDTD infer more flexible to different version of kg2. So in case the new xDTD database can't be completed before the relay. The code still works for kg2.8.0.1c during the relay.

**Comment by amykglen** [comment]
great, thanks @chunyuma!

**Comment by amykglen** [comment]
ok @edeutsch - I believe I've fixed that issue with the synonymizer and entity controller. please roll out `master` when you have a chance and I'll do more testing!

**Comment by edeutsch** [comment]
okay @amykglen , /test and kg2beta are synced with master and rolled out. They seem to work nicely, Please test!

**Comment by amykglen** [comment]
awesome, thanks! nice, and it even looks like the synonyms tab in the UI isn't angry! though there is probably some more info we could display there now - I'll write that up in an issue for Luis..

**Comment by amykglen** [comment]
@chunyuma also fixed a CRG issue, which I just merged into `master` - so we could use one more rollout when you have a chance @edeutsch..

**Comment by edeutsch** [comment]
done! test and kg2beta

**Comment by edeutsch** [comment]
okay, /test, /beta, /devED, /devLM, /kg2beta all seem good on master right now. fully rolled out as of now.
Please let me know if you see any problems.

**Comment by amykglen** [comment]
@acevedol - could you take care of this task for this rollout?

> - [ ] generate KGX files and upload them to the KGE Archive @acevedol

**Comment by amykglen** [comment]
per Lili we no longer need to generate KGX dumps for KG2c. I updated the rollout issue template to reflect that.

all five KG2 instances are now calling ITRB Plovers, so I just turned off our self-hosted KG2.8.0.1c Plover, and am closing this issue
